### ISSUE TITLE
didRefreshRegistrationToken method not updating to current APNS token after I tried a iOS method not recommended in the docs

### ISSUE THREAD

**Issue by reneeOlson** [firebase/quickstart-ios#343]
I had everything working great with the notifications since the 2.0.4 update to the FirebaseInstanceID.  I'm using cloud functions to handle the notification messages and I have method swizzling enabled. And when it was working I was using the method the Firebase docs recommended to use when a user taps on a notification  --- application(_:didReceiveRemoteNotification:fetchCompletionHandler:).  I was handling only data messages in the method --- application(received remoteMessage: MessagingRemoteMessage)

 But then I tested another method (maybe at the same time as above method) also with method swizzling enabled  -- the method that Firebase docs do not mention to use --- userNotificationCenter(_:willPresent:withCompletionHandler:)

It partially worked, as it provided the notification part of the message, but not the optional data payload.  I noticed that iOS seemed to of changed my device token for some reason and then the notifications and data messages stopped working suddenly. 

The cloud function logs show no more messages have been sent since then.  I noticed that the fcmToken does not contain the correct APNS token. It maintains the old one, despite me re-installing the app several times to get a new device token and I also removed the "willPresent" method and am only using the didReceiveRemoteNotification method now. It appears the automatic mapping does not update with the new APNS token.   

I'm completely stuck and not sure how to move forward.  Any suggestions?

Thanks!

**Comment by morganchen12** [comment]
IIRC FCM should revoke the token automatically if it tries to send a notification to an app that has been uninstalled. However, you can also manually reset the tokens via [`FIRInstanceID.deleteID(handler:)`](https://firebase.google.com/docs/reference/ios/firebaseinstanceid/api/reference/Classes/FIRInstanceID#/c:objc(cs)FIRInstanceID(im)deleteIDWithHandler:)

**Comment by reneeOlson** [comment]
The app was not uninstalled when I was sending the notifications. But I uninstalled the app to try and get firebase to reset the fcmToken to use the latest APNS token and it would not refresh it.  I will give that method a try and let you know if it works.  Thanks!

**Comment by reneeOlson** [comment]
I was not sure how to implement it and someone on stack overflow said they did it, but then they were not able to get notifications back again.  They implemented this code - 

`let instance = InstanceID.instanceID()
        _ = InstanceID.deleteID(instance)
        InstanceID.instanceID().deleteID { (err:Error?) in
            if err != nil{
                print(err.debugDescription);
            } else {
                print("Token Deleted");
            }
        }`

**Comment by reneeOlson** [comment]
also curious if this could be a bug.....if one uses that "willPresent notification" method with method swizzling enabled, which appears to break the automatic mapping for the new APNS token

**Comment by rsattar** [comment]
Hmm, I think something else is going on here, but FYI the APNs token logic is not affected by `willPresent` delegate method. [There *is* a fix to how that delegate method is swizzled](https://github.com/firebase/firebase-ios-sdk/pull/324), though, which is coming out in the next update.

**Comment by reneeOlson** [comment]
to clarify my issue, initially when the notifications were working, I was only using the method provided in the Firebase docs - application(_:didReceiveRemoteNotification:fetchCompletionHandler:).  I will edit this above to clarify.

Then I added the willPresent method. Maybe I was even using both at same time right before the notifications stopped working and I noticed the fcmToken will not refresh to latest APNS token. Could this perhaps caused a bug?

To fix right now, Morgan Chen says to use the FIRInstanceID.deleteID(handler:).  I pasted in an implementation above that someone did on stack overflow, but they said after they did it, the notifications did not work again.  So, was wondering what you guys think?  Any steps on how to implement it would be super helpful.  Thanks in advance!

**Comment by reneeOlson** [comment]
Now it seems to be working again!! So odd and I'm not sure why.  Sorry to have wasted your time.  I did have a question though, as I swear when I was verifying the fcmToken during my testing that the part after the colon was matching the deviceToken I got back in didRegisterForRemoteNotificationsWithDeviceToken.  And when the notifications stopped working, I noticed the part after the colon in the fcmToken no longer matched the deviceToken.  Was I wrong and the part after the colon does not match the deviceToken?
